common/util.go

```diff
@@ -80,6 +80,10 @@ const (
 	replicationServiceBusyMaxInterval        = 10 * time.Second
 	replicationServiceBusyExpirationInterval = 5 * time.Minute
 
+	taskCompleterInitialInterval    = 1 * time.Second
+	taskCompleterMaxInterval        = 10 * time.Second
+	taskCompleterExpirationInterval = 5 * time.Minute
```

Comment:
My biggest concerns with this are cases where the task doesn't ever complete because they lack pollers. If we didn't retry errTaskNotStarted then we'd be in a really aggressive loop of calling history over and over, so our only defense from this is the retry policy slowing down the subsequent calls. It looks like you've thought about this and the retry policy seems reasonable to address this.

For async dispatch though we operate with a timeout according to taskReader.getDispatchContext, which is typically 3s if zonal isolation is enabled or the user specified a tasklist rate limit. This means we'll just timeout and attempt again, rather than actually waiting 10 seconds between each request. With my upcoming changes for task leaking we'll have an even shorter timeout (zone specific) and then retries will have the longer one.

I think we probably should change task_reader to have a (much) longer timeout on the passive side. The current implementation doesn't make sense on the passive side at all, there's no reason to constantly timeout and then retry.

Reply:
Great point. I was going to check the timeout on the reader context but ended up forgetting to do it.
But I think we already have no timeouts in the standby cluster.
We set the context with timeout [here](https://github.com/cadence-workflow/cadence/blob/e9cf0fcab33697d48ae82eb3e6551e1f52584a62/service/matching/tasklist/task_reader.go#L408) after checking if it's active. If it's not, we return this [ctx](https://github.com/cadence-workflow/cadence/blob/e9cf0fcab33697d48ae82eb3e6551e1f52584a62/service/matching/tasklist/task_reader.go#L413) that is defined [here](https://github.com/cadence-workflow/cadence/blob/e9cf0fcab33697d48ae82eb3e6551e1f52584a62/service/matching/tasklist/task_reader.go#L95). It has cancellation but no timeout.
Also the taskListManager has itself a timeout. After a while, it shuts down. I tested it locally and the retries were aligned with the retryPolicy that I set.

Let me know if it was not super clear :)